lupa/_lupa.pyx

```diff
@@ -1597,13 +1602,40 @@ cdef int py_to_lua(LuaRuntime runtime, lua_State *L, object o, bint wrap_none=Fa
     elif isinstance(o, float):
         lua.lua_pushnumber(L, <lua.lua_Number><double>o)
         pushed_values_count = 1
+    elif isinstance(o, _PyProtocolWrapper):
+        type_flags = (<_PyProtocolWrapper> o)._type_flags
+        o = (<_PyProtocolWrapper> o)._obj
+        pushed_values_count = py_to_lua_custom(runtime, L, o, type_flags)
+    elif recursive and isinstance(o, (list, Sequence)):
+        if mapped_objs is None:
+            mapped_objs = {}
+        if id(o) not in mapped_objs:
+            lua.lua_createtable(L, _len_as_int(len(o)), 0)   # create a table at the top of stack, with narr already known
+            mapped_objs[id(o)] = lua.lua_gettop(L)
+            for i, v in enumerate(o, 1):
+                py_to_lua(runtime, L, v, wrap_none, recursive, mapped_objs)
+                lua.lua_rawseti(L, -2, i)
+        else: # self-reference detected
+            idx = mapped_objs[id(o)]
+            lua.lua_pushvalue(L, idx)
+        pushed_values_count = 1
+    elif recursive and isinstance(o, (dict, Mapping)):
+        if mapped_objs is None:
+            mapped_objs = {}
+        if id(o) not in mapped_objs:
+            lua.lua_createtable(L, 0, _len_as_int(len(o)))  # create a table at the top of stack, with nrec already known
+            mapped_objs[id(o)] = lua.lua_gettop(L)
+            for key, value in o.items():
+                py_to_lua(runtime, L, key, wrap_none, recursive, mapped_objs)
+                py_to_lua(runtime, L, value, wrap_none, recursive, mapped_objs)
+                lua.lua_rawset(L, -3)
+        else: # self-reference detected
+            idx = mapped_objs[id(o)]
+            lua.lua_pushvalue(L, idx)
+        pushed_values_count = 1
```

Comment:
I'd reverse the conditions to put the simple "already mapped" case first.
```suggestion
    elif recursive and isinstance(o, (list, Sequence)):
        if mapped_objs is None:
            mapped_objs = {}
        if id(o) in mapped_objs:
            obj = mapped_objs[id(o)]
            lua.lua_pushvalue(L, obj)
        else:
            lua.lua_createtable(L, _len_as_int(len(o)), 0)  # create a table at the top of stack, with narr already known
            mapped_objs[id(o)] = lua.lua_gettop(L)
            for i, v in enumerate(o, 1):
                py_to_lua(runtime, L, v, wrap_none, recursive, mapped_objs)
                lua.lua_rawseti(L, -2, i)
        pushed_values_count = 1
    elif recursive and isinstance(o, (dict, Mapping)):
        if mapped_objs is None:
            mapped_objs = {}
        if id(o) in mapped_objs:
            obj = mapped_objs[id(o)]
            lua.lua_pushvalue(L, obj)
            lua.lua_createtable(L, 0, _len_as_int(len(o)))  # create a table at the top of stack, with nrec already known
            mapped_objs[id(o)] = lua.lua_gettop(L)
            for key, value in o.items():
                py_to_lua(runtime, L, key, wrap_none, recursive, mapped_objs)
                py_to_lua(runtime, L, value, wrap_none, recursive, mapped_objs)
                lua.lua_rawset(L, -3)
        pushed_values_count = 1
```
Actually, I'd even reuse the existing `LuaRuntime.table_from()` implementation here. Or, rather, extract a `cdef` function from it and use it in both places. This seems needlessly duplicated code.